### PR TITLE
Adds saslprep and SCRAM-SHA-256

### PR DESCRIPTION
**Pull request by daprahamian**
No longer WIP.

Includes SCRAM-SHA-256, saslprep for Node >= 6, and Mechanism Negotiation.

Fixes NODE-1303
Fixes NODE-1311

### CODE REVIEW

File: lib/auth/scram.js [hunk 1]
```diff
@@ -124,6 +155,26 @@ ScramSHA1.prototype.auth = function(server, connections, db, username, password,
   var numberOfValidConnections = 0;
   var errorObject = null;
 
+  let cryptoMethod = 'sha1';
+  let mechanism = 'SCRAM-SHA-1';
+  let processedPassword;
+
+  if (this.cryptoMethod === 'sha256') {
+    cryptoMethod = 'sha256';
```
Comment (mbroadst) [inline lib/auth/scram.js]: can't you just use `this.cryptoMethod` later rather than creating a new variable?

File: lib/auth/scram.js [hunk 2]
```diff
@@ -132,13 +183,21 @@ ScramSHA1.prototype.auth = function(server, connections, db, username, password,
     // Create a random nonce
     var nonce = crypto.randomBytes(24).toString('base64');
     // var nonce = 'MsQUY9iw0T9fx2MUEz6LZPwGuhVvWAhc'
-    var firstBare = f('n=%s,r=%s', username, nonce);
+
+    // NOTE: This is done b/c Javascript uses UTF-16, but the server is hashing in UTF-8.
+    // Since the username is not sasl-prep-d, we need to do this here.
+    const firstBare = Buffer.concat([
```
Comment (mbroadst) [inline lib/auth/scram.js]: do we want to further explicitly state the encoding as `utf-8` even though that's the default?

File: lib/topologies/server.js [hunk 3]
```diff
@@ -21,6 +21,57 @@ var inherits = require('util').inherits,
   SessionMixins = require('./shared').SessionMixins,
   relayEvents = require('./shared').relayEvents;
 
+function getSaslSupportedMechs(options) {
+  if (!options) {
+    return {};
+  }
+
+  const authArray = options.auth || [];
+  const authMechanism = authArray[0] || options.authMechanism;
```
Comment (mbroadst) [inline lib/topologies/server.js]: This looks like it can still potentially be `undefined`

File: lib/topologies/server.js [hunk 4]
```diff
@@ -21,6 +21,57 @@ var inherits = require('util').inherits,
...
+    if (ismaster.maxWireVersion >= 3) {
+      return 'scram-sha-1';
+    }
+  }
+
+  // Default for wireprotocol < 3
+  return 'mongocr';
+}
+
+function parseOutIsMasterError(err, result) {
```
Comment (mbroadst) [inline lib/topologies/server.js]: nit: is it really parsing, or extracting?

File: test/tests/functional/basic_single_server_auth_tests.js [hunk 5]
```diff
@@ -329,7 +381,7 @@ describe.skip('Basic single server auth tests', function() {
     }
   });
 
-  it('should correctly authenticate server using scram-sha-1 using connect auth then logout', {
+  it.skip('should correctly authenticate server using scram-sha-1 using connect auth then logout', {
```
Comment (mbroadst) [inline test/tests/functional/basic_single_server_auth_tests.js]: what's going on here? should we perhaps skip it only for <3.7.3?

File: test/tests/functional/basic_single_server_auth_tests.js [hunk 6]
```diff
@@ -8,7 +8,59 @@ var expect = require('chai').expect,
...
+          bson: new Bson()
+        });
+
+        server.on('connect', _server => {
+          dropUser((dropUserErr, dropUserRes) => {
+            expect(dropUserErr).to.be.null;
+            expect(dropUserRes).to.exist;
+
+            _server.destroy({ force: true });
+            expect(Object.keys(Connection.connections()).length).to.equal(0);
```
Comment (mbroadst) [inline test/tests/functional/basic_single_server_auth_tests.js]: I've been trying to remove all traces of these types of tests. This was here to ensure that all connections were really closed before moving on to the next test, and we have better ways of doing this now with the `ConnectionSpy` (at least). Ideally, `destroy` would take a callback but I just made [a ticket](https://jira.mongodb.org/browse/NODE-1443) for that

File: lib/topologies/server.js (hunk 3 above)
Comment (daprahamian) [reply lib/topologies/server.js]: It is possible for it to be `undefined`. That's one of the cases that we want to handle with the `saslSupportedMechs`

File: lib/auth/scram.js (hunk 1 above)
Comment (daprahamian) [reply lib/auth/scram.js]: I had something in mind about decoupling the symbol we use from the nodejs value. If we wanted to switch to `Symbol`s for example. But you're right, it isn't necessary.

File: lib/auth/scram.js (hunk 2 above)
Comment (daprahamian) [reply lib/auth/scram.js]: Yeah, I can add that in.

File: test/tests/functional/basic_single_server_auth_tests.js (hunk 5 above)
Comment (daprahamian) [reply test/tests/functional/basic_single_server_auth_tests.js]: This  test is just broken in general, before I even made changes.

File: test/tests/functional/basic_single_server_auth_tests.js (hunk 6 above)
Comment (daprahamian) [reply test/tests/functional/basic_single_server_auth_tests.js]: Ok, I'll remove it.

File: lib/auth/scram.js [hunk 7]
```diff
@@ -124,6 +155,25 @@ ScramSHA1.prototype.auth = function(server, connections, db, username, password,
   var numberOfValidConnections = 0;
   var errorObject = null;
 
+  let cryptoMethod = this.cryptoMethod;
```
Comment (mbroadst) [inline lib/auth/scram.js]: wow I thought this was unused, but this method is like hundreds of lines long (!)

File: lib/auth/scram.js [hunk 8]
```diff
@@ -124,6 +155,25 @@ ScramSHA1.prototype.auth = function(server, connections, db, username, password,
...
+  let mechanism = 'SCRAM-SHA-1';
+  let processedPassword;
+
+  if (this.cryptoMethod === 'sha256') {
```
Comment (mbroadst) [inline lib/auth/scram.js]: nit: you can definitely reuse `cryptoMethod` here though

File: lib/topologies/mongos.js [hunk 9]
```diff
@@ -37,12 +37,7 @@ const BSON = retrieveBSON();
  * server.connect();
  */
 
-var MongoCR = require('../auth/mongocr'),
-  X509 = require('../auth/x509'),
-  Plain = require('../auth/plain'),
-  GSSAPI = require('../auth/gssapi'),
-  SSPI = require('../auth/sspi'),
-  ScramSHA1 = require('../auth/scram');
+const defaultAuthProviders = require('../auth/defaultAuthProviders').defaultAuthProviders;
```
Comment (mbroadst) [inline lib/topologies/mongos.js]: love these cleanups!

File: package.json [hunk 10]
```diff
@@ -44,5 +44,8 @@
   "bugs": {
     "url": "https://github.com/mongodb-js/mongodb-core/issues"
   },
-  "homepage": "https://github.com/mongodb-js/mongodb-core"
+  "homepage": "https://github.com/mongodb-js/mongodb-core",
+  "optionalDependencies": {
```
Comment (mbroadst) [inline package.json]: should we switch all of our `peerOptionalDependencies` to just be `optionalDependencies` 🤔

Comment (mbroadst) [reply package.json]: I just went and saw a number of issues users were having with `require_optional` (e.g. [this](https://github.com/christkv/require_optional/issues/3)), maybe we should discuss removing it in favor of just `optionalDependencies`

**Comment by daprahamian** [comment]
PR Updated with feedback

**Comment by mbroadst** [comment]
:shipit:

**Pull request merged**